

// The following messages are used in GRR's web API.

syntax = "proto2";

import "grr/proto/flows.proto";
import "grr/proto/jobs.proto";
import "grr/proto/knowledge_base.proto";
import "grr/proto/output_plugin.proto";
import "grr/proto/semantic.proto";
import "grr/proto/sysinfo.proto";



//
// API call handlers arguments (see gui/api_call_handlers.py).
//

message ApiAuthorization {
  optional string handler = 1 [(sem_type) = {
      description: "api_call_handlers.ApiCallHandler class name that this "
                   "authorization applies to."
    }];
  repeated string users = 2 [(sem_type) = {
      description: "List of users allowed to call the API."
    }];
  repeated string groups = 3[(sem_type) = {
      description: "List of groups allowed to call the API."
    }];
}

message ApiCallAdditionalArgs {
  optional string name = 1;
  optional string type = 2;
  optional bytes args = 3 [(sem_type) = {
      dynamic_type: "GetArgsClass"
    }];
}

message ApiGetAff4ObjectArgs {
  optional string aff4_path = 1 [(sem_type) = {
      type: "RDFURN",
      description: "Path of the object to be rendered."
    }];

  repeated ApiCallAdditionalArgs additional_args = 2;
}

message ApiGetAff4IndexArgs {
  optional string aff4_path = 1 [(sem_type) = {
      type: "RDFURN",
      description: "Path of the object to be rendered."
    }];
}

message ApiListClientsArgs {
  optional string query = 1 [(sem_type) = {
      description: "Search query string. See here for syntax: "
      "https://github.com/google/grr-doc/blob/master/"
      "user_manual.adoc#searching-for-a-client"
  }];
  optional int64 offset = 2 [(sem_type) = {
      description: "Found clients starting offset."
    }];
  optional int64 count = 3 [(sem_type) = {
      description: "Number of found client to fetch."
    }];
}

message ApiListClientsResult {
  repeated ApiClient items = 1;
}

message ApiGetClientArgs {
  optional string client_id = 1 [(sem_type) = {
      type: "ClientURN",
      description: "Client id."
    }];
}

message ApiCreateFlowArgs {
  optional string client_id = 1 [(sem_type) = {
      type: "ClientURN",
      description: "Client id."
    }];
  optional ApiFlow flow = 2;
}

message ApiCancelFlowArgs {
  optional string client_id = 1 [(sem_type) = {
      type: "ClientURN",
      description: "Client id."
    }];
  optional string flow_id = 2 [(sem_type) = {
      type: "SessionID",
      description: "Flow id."
    }];
}

message ApiGetFlowStatusArgs {
  optional string client_id = 1 [(sem_type) = {
      type: "ClientURN",
      description: "Client id."
    }];
  optional string flow_id = 2 [(sem_type) = {
      type: "SessionID",
      description: "Flow id."
    }];
}

message ApiListFlowResultsArgs {
  optional string client_id = 1 [(sem_type) = {
      type: "ClientURN",
      description: "Client id."
    }];
  optional string flow_id = 2 [(sem_type) = {
      type: "SessionID",
      description: "Flow id."
    }];
  optional int64 offset = 3 [(sem_type) = {
      description: "Results items starting offset."
    }];
  optional int64 count = 4 [(sem_type) = {
      description: "Max number of results to fetch."
    }];
  optional string filter = 5 [(sem_type) = {
      description: "Return only results whose string representation "
      "contains given substring."
    }];
}

message ApiGetFlowResultsExportCommandArgs {
  optional string client_id = 1 [(sem_type) = {
      type: "ClientURN",
      description: "Client id."
    }];
  optional string flow_id = 2 [(sem_type) = {
      type: "SessionID",
      description: "Flow id."
    }];
}

message ApiListFlowOutputPluginsArgs {
  optional string client_id = 1 [(sem_type) = {
      type: "ClientURN",
      description: "Client id."
    }];
  optional string flow_id = 2 [(sem_type) = {
      type: "SessionID",
      description: "Flow id."
    }];
}

message ApiListFlowOutputPluginsResult {
  repeated ApiOutputPlugin items = 2;
}

message ApiListFlowOutputPluginLogsArgs {
  optional string client_id = 1 [(sem_type) = {
      type: "ClientURN",
      description: "Client id."
    }];
  optional string flow_id = 2 [(sem_type) = {
      type: "SessionID",
      description: "Flow id."
    }];
  optional string plugin_id = 3 [(sem_type) = {
      description: "Output plugin id."
    }];

  optional int64 offset = 4 [(sem_type) = {
      description: "Logs starting offset."
    }];
  optional int64 count = 5 [(sem_type) = {
      description: "Max number of log entries."
    }];
}

message ApiListFlowOutputPluginErrorsArgs {
  optional string client_id = 1 [(sem_type) = {
      type: "ClientURN",
      description: "Client id."
    }];
  optional string flow_id = 2 [(sem_type) = {
      type: "SessionID",
      description: "Flow id."
    }];
  optional string plugin_id = 3 [(sem_type) = {
      description: "Output plugin id."
    }];

  optional int64 offset = 4 [(sem_type) = {
      description: "Errors starting offset."
    }];
  optional int64 count = 5 [(sem_type) = {
      description: "Max number of errors."
    }];
}

message ApiArchiveFlowFilesArgs {
  optional string client_id = 1 [(sem_type) = {
      type: "ClientURN",
      description: "Client id."
    }];
  optional string flow_id = 2 [(sem_type) = {
      description: "Flow id.",
      type: "SessionID"
    }];
  optional ExportCollectionFilesAsArchiveArgs.ArchiveFormat archive_format = 3;
};

message ApiListFlowDescriptorsArgs {
  enum FlowType {
    CLIENT = 0;
    GLOBAL = 1;
  };
  optional FlowType flow_type = 1 [(sem_type) = {
      description: "If specified, narrows the list to the given type of flows "
      "only."
    }];
};

message ApiListClientFlowsArgs {
  optional string client_id = 1 [(sem_type) = {
      type: "ClientURN",
      description: "Client id."
    }];
  optional int64 offset = 2 [(sem_type) = {
      description: "Starting offset."
    }];
  optional int64 count = 3 [(sem_type) = {
      description: "Max number of flows to fetch."
    }];
};

message ApiListClientFlowsResult {
  repeated ApiFlow items = 1;
};

message ApiAddClientsLabelsArgs {
  repeated string client_ids = 1 [(sem_type) = {
      type: "ClientURN",
      description: "Client id."
    }];
  repeated string labels = 2;
}

message ApiRemoveClientsLabelsArgs {
  repeated string client_ids = 1 [(sem_type) = {
      type: "ClientURN",
      description: "Client id."
    }];
  repeated string labels = 2;
}

message ApiListClientsLabelsResult {
  repeated AFF4ObjectLabel items = 1;
}

message ApiListKbFieldsResult {
  repeated string items = 2;
}

message ApiListHuntsArgs {
  optional int64 offset = 1 [(sem_type) = {
      description: "Starting offset."
    }];
  optional int64 count = 2 [(sem_type) = {
      description: "Max number of items to fetch."
    }];

  // Filtering criterias.
  optional string created_by = 3 [(sem_type) = {
      description: "Only return hunts created by a given user. "
      "If approved_by or/and description_contains are also supplied, "
      "then logical AND is applied to all the criterias. NOTE: this filter "
      "can only be used in conjunction with 'active_within' filter (to "
      "prevent queries of death)."
    }];
  optional string description_contains = 4 [(sem_type) = {
      description: "Only return hunts where description contains given "
      "substring (matching is case-insensitive)."
      "If created_by or/and approved_by are also supplied, "
      "then logical AND is applied to all the criterias. NOTE: this filter "
      "can only be used in conjunction with 'active_within' filter (to "
      "prevent queries of death)."
    }];
  optional uint64 active_within = 5 [(sem_type) = {
      type: "Duration",
      description: "Only return hunts that were active within given time "
      "duration."
    }];
}

message ApiGetHuntArgs {
  optional string hunt_id = 1 [(sem_type) = {
      description: "Hunt id."
    }];
}

message ApiListHuntLogsArgs {
  optional string hunt_id = 1 [(sem_type) = {
      description: "Hunt id."
    }];
  optional int64 offset = 2 [(sem_type) = {
      description: "Log items starting offset."
    }];
  optional int64 count = 3 [(sem_type) = {
      description: "Max number of log items to fetch."
    }];
};

message ApiListHuntErrorsArgs {
  optional string hunt_id = 1 [(sem_type) = {
      description: "Hunt id."
    }];
  optional int64 offset = 2 [(sem_type) = {
      description: "Hunt errors starting offset."
    }];
  optional int64 count = 3 [(sem_type) = {
      description: "Max number of errors to fetch."
    }];
};

message ApiListHuntResultsArgs {
  optional string hunt_id = 1 [(sem_type) = {
      description: "Hunt id."
    }];
  optional int64 offset = 2 [(sem_type) = {
      description: "Hunt results starting offset."
    }];
  optional int64 count = 3 [(sem_type) = {
      description: "Max number of results to fetch."
    }];
  optional string filter = 4 [(sem_type) = {
      description: "Return only results whose string representation "
      "contains given substring."
    }];
};

message ApiGetHuntResultsExportCommandArgs {
  optional string hunt_id = 1 [(sem_type) = {
      description: "Hunt id."
    }];
};

message ApiListHuntOutputPluginsArgs {
  optional string hunt_id = 1 [(sem_type) = {
      description: "Hunt id."
    }];
};

message ApiListHuntOutputPluginLogsArgs {
  optional string hunt_id = 1 [(sem_type) = {
      description: "Hunt id."
    }];
  optional string plugin_id = 2 [(sem_type) = {
      description: "Plugin id."
    }];

  optional int64 offset = 3 [(sem_type) = {
      description: "Status messages offset."
    }];
  optional int64 count = 4 [(sem_type) = {
      description: "Number of status messages to fetch."
    }];
};

message ApiListHuntOutputPluginErrorsArgs {
  optional string hunt_id = 1 [(sem_type) = {
      description: "Hunt id."
    }];
  optional string plugin_id = 2 [(sem_type) = {
      description: "Plugin id."
    }];

  optional int64 offset = 3 [(sem_type) = {
      description: "Errors offset."
    }];
  optional int64 count = 4 [(sem_type) = {
      description: "Number of errors messages to fetch."
    }];
};

message ApiArchiveHuntFilesArgs {
  optional string hunt_id = 1 [(sem_type) = {
      description: "Hunt id.",
      type: "SessionID"
    }];
  optional ExportCollectionFilesAsArchiveArgs.ArchiveFormat archive_format = 2;
};

message ApiCreateHuntArgs {
  optional HuntRunnerArgs hunt_runner_args = 1 [(sem_type) = {
      description: "Hunt runner arguments."
    }];
  optional GenericHuntArgs hunt_args = 2 [(sem_type) = {
      description: "Hunt arguments."
    }];
};

message ApiGetRDFValueDescriptorArgs {
  optional string type = 1 [(sem_type) = {
      description: "RDFValue type name."
    }];
}

message ApiListStatsStoreMetricsMetadataArgs {
  enum Component {
    WORKER = 0;
    FRONTEND = 1;
  }
  optional Component component = 1;
}

message ApiListCronJobsArgs {
  optional int64 offset = 1 [(sem_type) = {
      description: "Starting offset."
    }];
  optional int64 count = 2 [(sem_type) = {
      description: "Max number of cron jobs to fetch."
    }];
};


// Next field number: 8
message ApiGetStatsStoreMetricArgs {
  enum Component {
    WORKER = 0;
    FRONTEND = 1;
  }
  optional Component component = 1;
  optional string metric_name = 2;

  optional uint64 start = 3 [(sem_type) = {
      type: "RDFDatetime",
      description: "Fetch metrics data starting from this timestamp. If "
      "not set, will default to 1 hour ago."
    }];
  optional uint64 end = 4 [(sem_type) = {
      type: "RDFDatetime",
      description: "Fetch metrics data from the interval ending with "
      "this timestamp. If not set, will default to the current time."
    }];

  optional uint64 rate = 7 [(sem_type) = {
      type: "Duration",
      description: "If set, metric's rate over given window will be returned "
      "instead of plain values."
    }];

  enum DistributionHandlingMode {
    DH_SUM = 0 [(description) = "Sum of all values recorded in a distribution "
             "will be fetched."];
    DH_COUNT = 1 [(description) = "Number of all values recorded in a distribution "
               "will be fetched."];
  };
  optional DistributionHandlingMode distribution_handling_mode = 5 [(sem_type) = {
      description: "Determines which values to fetch when dealing with distributions."
    }];

  enum AggregationMode {
    AGG_SUM = 0 [(description) = "Values will be summed."];
    AGG_MEAN = 1 [(description) = "Mean value will be calculated."];
    AGG_NONE = 2 [(description) = "No aggregation will be performed."];
  };
  optional AggregationMode aggregation_mode = 6 [(sem_type) = {
      description: "Aggregation method. As metrics are usually collected across "
        "multiple processes, most of the time they have to be aggregated."
    }];
}

message ApiListArtifactsArgs {
  optional int64 offset = 1 [(sem_type) = {
      description: "Starting offset."
    }];
  optional int64 count = 2 [(sem_type) = {
      description: "Max number of items to fetch."
    }];
}

message ApiUploadArtifactArgs {
  optional bytes artifact = 1 [(sem_type) = {
      description: "Artifact YAML definition."
    }];
}

message ApiDeleteArtifactsArgs {
  repeated string names = 1 [(sem_type) = {
      description: "Names of the artifacts to be deleted."
    }];
}

message ApiGetConfigOptionArgs {
  optional string name = 1 [(sem_type) = {
      description: "Name of the config option to fetch."
    }];
}

message ApiListUserHuntApprovalsArgs {
  optional int64 offset = 1 [(sem_type) = {
      description: "Approvals starting offset."
    }];
  optional int64 count = 2 [(sem_type) = {
      description: "Number of approvals to fetch."
    }];
}

message ApiListUserCronApprovalsArgs {
  optional int64 offset = 1 [(sem_type) = {
      description: "Approvals starting offset."
    }];
  optional int64 count = 2 [(sem_type) = {
      description: "Number of approvals to fetch."
    }];
}

message ApiListUserClientApprovalsArgs {
  optional int64 offset = 1 [(sem_type) = {
      description: "Approvals starting offset."
    }];
  optional int64 count = 2 [(sem_type) = {
      description: "Number of approvals to fetch."
    }];
}

message ApiListUserClientApprovalsResult {
  repeated ApiUserClientApproval items = 1;
}

message ApiStartGetFileOperationArgs {
  optional string hostname = 1 [(sem_type) = {
      description: "Target host. Can be FQDN or client ID."
    }];
  repeated string paths = 2 [(sem_type) = {
      type: "GlobExpression",
      description: "Fetch this number of elements."
    }];
  optional uint64 max_file_size = 3 [(sem_type) = {
      type: "ByteSize",
      description: "Maximum size of file we will download.",
    }, default=20000000]; // 20MB
  optional FileFinderAction.Action action = 4 [(sem_type) = {
      description: "Action to apply to found files"
    }, default=DOWNLOAD];
  optional PathSpec.PathType pathtype = 5 [(sem_type) = {
      description: "Path type to glob in.",
    }, default=OS];
}

//
// AFF4 object renderers arguments (see gui/api_aff4_object_renderers.py).
//
message ApiAFF4ObjectRendererArgs {
  optional int64 limit_lists = 1 [(sem_type) = {
      description: "If >0, lists of values in the response will contain "
      "no more than specified number of items. This is an effective way to "
      "protect the UI from fetching and presenting enormously big datasets. "
      "If 0, no lists will be rendered at all. If -1, lists will be rendered "
      "in their entirety."
    }, default = -1];
}

message ApiGRRHuntRendererArgs {
  optional bool with_full_summary = 1 [(sem_type) = {
      description: "If True, render stats, clients counts, rules and args."
    }];
}

message ApiRDFValueCollectionRendererArgs {
  optional int64 offset = 1 [(sem_type) = {
      description: "Fetch items starting from this offset."
    }];
  optional int64 count = 2 [(sem_type) = {
      description: "Fetch this number of elements."
    }];
  optional bool with_total_count = 3 [(sem_type) = {
      description: "Include total number of elements in the collection into the"
        " response. NOTE: for some collections this may be expensive."
    }];
  optional string filter = 4 [(sem_type) = {
      description: "Return only collection items whose string representation "
      "contains given substring."
    }];

  optional int64 items_limit_lists = 5 [(sem_type) = {
      description: "If >0, lists of values in the response will contain "
      "no more than specified number of items. This is an effective way to "
      "protect the UI from fetching and presenting enormously big datasets. "
      "If 0, no lists will be rendered at all. If -1, lists will be rendered "
      "in their entirety."
    }, default = -1];
}


// API entities. These are uses in API calls responses.

message ApiUserClientApproval {
  optional ApiClient subject = 1;

  optional string reason = 2;
  optional bool is_valid = 3;
  optional string is_valid_message = 4;
}

message ApiClient {
  optional string urn = 1 [(sem_type) = {
      type: "ClientURN",
      description: "Client URN."
    }];

  optional ClientInformation agent_info = 2;
  optional HardwareInfo hardware_info = 3;
  optional Uname os_info = 4;

  optional uint64 first_seen_at = 6 [(sem_type) = {
      type: "RDFDatetime",
      description: "When flow was created."
    }];
  optional uint64 last_seen_at = 7 [(sem_type) = {
      type: "RDFDatetime",
      description: "Last time when client checked in."
    }];
  optional uint64 last_booted_at = 8 [(sem_type) = {
      type: "RDFDatetime",
      description: "Last boot time."
    }];
  optional uint64 last_clock = 9 [(sem_type) = {
      type: "RDFDatetime",
      description: "Client clocks value during latest checkin."
    }];
  optional uint64 last_crash_at = 10 [(sem_type) = {
      type: "RDFDatetime",
      description: "Last crash time."
    }];

  repeated AFF4ObjectLabel labels = 11;
  repeated Interface interfaces = 12;
  repeated User users = 13;
  repeated Volume volumes = 14;
}

message ApiFlow {
  // Enum values here correspond to Flow.State values.
  enum State {
    RUNNING = 0;
    TERMINATED = 1;
    ERROR = 3;
    CLIENT_CRASHED = 4;
  };

  optional string urn = 1 [(sem_type) = {
      type: "SessionID",
      description: "Flow URN."
    }];
  optional string name = 2 [(sem_type) = {
      description: "Flow name.",
    }];
  optional AnyValue args = 3 [(sem_type) = {
      description: "Flow arguments.",
      dynamic_type: "GetArgsClass"
    }];
  optional FlowRunnerArgs runner_args = 4 [(sem_type) = {
      description: "Flow runner arguments."
    }];
  optional State state = 5 [(sem_type) = {
      description: "Current flow state."
    }];
  optional uint64 started_at = 6 [(sem_type) = {
      type: "RDFDatetime",
      description: "When flow was created."
    }];
  optional uint64 last_active_at = 7 [(sem_type) = {
      type: "RDFDatetime",
      description: "When flow was last active."
    }];
  optional string creator = 8 [(sem_type) = {
      description: "Who started the flow."
    }];

  repeated ApiFlow nested_flows = 9 [(sem_type) = {
      description: "Nested flows started by this flow."
    }];
}

message ApiCronJob {
  enum State {
    ENABLED = 0;
    DISABLED = 1;
  };

  optional string urn = 1 [(sem_type) = {
      type: "RDFURN",
      description: "Cron job URN."
    }];
  optional string description = 2 [(sem_type) = {
      description: "Cron job description."
    }];
  optional string flow_name = 3 [(sem_type) = {
      description: "Cron flow name."
    }];
  optional AnyValue flow_args = 4 [(sem_type) = {
      description: "Cron flow arguments.",
      dynamic_type: "GetArgsClass"
    }];
  optional FlowRunnerArgs flow_runner_args = 5 [(sem_type) = {
      description: "Flow runner arguments."
    }];
  optional uint64 periodicity = 6 [(sem_type) = {
      type: "Duration",
      description: "Interval between cron runs."
    }];
  optional uint64 lifetime = 7 [(sem_type) = {
      type: "Duration",
      description: "How long each run of the cron should be allowed to run. "
      "Runs that exceed lifetime will be killed. This is complimentary but"
      " different to periodicity e.g. it allows us to run jobs weekly that"
      " should be killed if they take more than a few hours.",
    }];
  optional bool allow_overruns = 8 [(sem_type) = {
      description: "If true, don't kill the previous run if new one starts.",
      label: ADVANCED
    }];

  optional State state = 9 [(sem_type) = {
      description: "Current flow state."
    }];
  optional uint64 last_run_time = 10 [(sem_type) = {
      type: "RDFDatetime",
      description: "Last time when this cron job ran."
    }];
  optional bool is_failing = 11 [(sem_type) = {
      description: "Is this cron job constantly failing?"
    }];
}

message ApiOutputPlugin {
  optional string id = 1 [(sem_type) = {
      description: "String identifying the output plugin."
    }];
  optional OutputPluginDescriptor plugin_descriptor = 2 [(sem_type) = {
      description: "Descriptor of the output plugin."
    }];
  optional AnyValue state = 3 [(sem_type) = {
      description: "State of the plugin.",
      dynamic_type: "GetStateClass"
    }];
  optional OutputPluginVerificationResult last_verification_result = 4 [(sem_type) = {
      description: "Results of the last verification check done on the output "
      "of this plugin. Verification checks check that expected output exists "
      "and that it's complete. Not all output plugins support verification. "
      "For example CSV and Email plugins output can't be verified. On the "
      "other hand, BigQuery output can be verified, as BigQuery service can "
      "be queried for the results that were supposed to be uploaded."
    }];
}

message ApiVfsRefreshOperation {
  optional string client_id = 1 [(sem_type) = {
      type: "ClientURN",
      description: "Client id."
    }];
  optional string vfs_path = 2 [(sem_type) = {
      description: "AFF4 path of a VFS folder to be refreshed."
    }];
  optional uint64 max_depth = 3 [(sem_type) = {
      description: "Maximum recursion depth.",
    }];
}
